File: 06/Java/TripleExpression/BinaryOperator.java
package TripleExpression;

abstract public class BinaryOperator implements TripleExpression {
    protected TripleExpression fOp;
    protected TripleExpression sOp;
    
    BinaryOperator(TripleExpression first, TripleExpression second) {
        fOp = first;
        sOp = second;
    }

    abstract protected int apply(int a, int b);
    
    public int evaluate(int x, int y, int z) {
        return apply(fOp.evaluate(x, y, z), sOp.evaluate(x, y, z));
    }
}

File: 07/CSharp/TripleExpressionApp/TripleExpressionApp/BinaryOperator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TripleExpressionApp
{
    abstract public class BinaryOperator : ITripleExpression
    {
        protected ITripleExpression fOp;
        protected ITripleExpression sOp;

        abstract protected double Apply(double a, double b);
        public double Evaluate(double x, double y, double z)
        {
            return Apply(fOp.Evaluate(x, y, z), sOp.Evaluate(x, y, z));
        }
    }
}
